test_app/app/channels/application_cable/connection.rb

```diff
@@ -0,0 +1,4 @@
+module ApplicationCable
```

Comment:
What is this file about?

Reply:
# Connection.rb

## What is **WebSocket**

WebSocket is a computer communications protocol, providing full-duplex communication channels over a single TCP connection. 

### HTTP Request

Unlike our usual HTTP request where the client sends an HTTP or HTTPS request, and the server sends the response. After that, the whole connection got terminated, so each of the requests is a new connection between client and server and also independent of each other ( one request does not contain info of other request and no order ).

To observe the protocol request and response process, here I tried a ruby code as below to send an HTTP request to the server.

```ruby
require 'socket'

body = <<~EOF
    GET / HTTP/1.0\r
    Host: localhost:3000\r
    \r
  EOF

Socket.tcp("localhost", 3000){|sock|
    sock.print body
    sock.close_write
    puts sock.read
}
```

Ruby use rack to handle the http. It takes in an env parameter which contains all the information about the request ( More details about it avalaible at  [https://github.com/rack/rack/blob/main/SPEC.rdoc](https://github.com/rack/rack/blob/main/SPEC.rdoc) ). And the server respond with the header that consist of status code ( 200 in below example ), some headers ( ‘Content-Type’ in below example ) and the response body.

```ruby
require 'rack'

app = Proc.new do |env|
    puts env
    [
        '200',
        {'Content-Type' => 'text/html'},
        ['Response Body']
    ]
end

Rack::Handler::WEBrick.run app
```

After the server responds with a 200 OK status, the below is what we will receive from the sock print body where we put in send http request code and now the connection is terminated.

```html
HTTP/1.0 200 OK

Content-Type: text/html; charset=utf-8

Cache-Control: no-store, must-revalidate, private, max-age=0

Content-Security-Policy: default-src 'self'; img-src 'self' data:; script-src 'self' 'unsafe-inline'; style-src 'self' 'unsafe-inline'

X-Request-Id: a8762ebe-4e95-4dbc-9cb6-f59463698b2c

X-Runtime: 0.075966

X-Frame-Options: sameorigin

X-Content-Type-Options: nosniff

X-XSS-Protection: 1; mode=block

X-Download-Options: noopen

X-Permitted-Cross-Domain-Policies: none

X-MiniProfiler-Original-Cache-Control: max-age=0, private, must-revalidate

X-MiniProfiler-Ids: pbjo13gn4nckxd0sbj0k,ie2q5hpg7eihvq4mf048,oqlpizkmf3g2vcu7xv1r,eath0vl655u82ier3d6d,tecy7c9vi08dblpujmrl,nu9g7iambtbx4g88yhtc

Set-Cookie: __profilin=p%3Dt; path=/; HttpOnly; SameSite=Lax

Content-Length: 400598
```

### Websocket

Websocket is bidirectional, a full-duplex protocol that is used particularly for client-server communication. It is a stateful protocol, which means the connection between client and server will keep alive until it is terminated by either party (client or server).

Again, to observe the process, here I also tried with a ruby code to send a request to the server to ask for an upgrade from normal HTTP to Websocket. Before it returns status 200 ok, we now need the server to first return a 101 switch protocol. When we have that response, we will have this magic conversion of the get response to some sort of bi-directional socket.

```ruby
require 'socket'

body = <<~EOF
    GET ws://localhost:3000/cable HTTP/1.1\r
    Host: localhost:3000\r
    Connection: Upgrade\r
    Upgrade: websocket\r
    Origin: http://localhost:3000\r
    \r
  EOF

Socket.tcp("localhost", 3000){|sock|
    sock.print body
    sock.close_write
    puts sock.read
}
```

In the previous HTTP section, we explain the Ruby use Rack to handle HTTP request. How it happens with Websocket then?

The rack is using multiple Proc to handle the HTTP request, the Proc keeps calling the next Proc and passing in the env parameter until it reaches the last Proc. After that, it will start to return with the Server response. 

OK so that is how our rack handles typical http request in more detail. To create a websocket connection, we have to somehow drag the client socket out and perform some operations on it, and we can do it by using hack hijack. The below code shows how the websocket connection is established by calling rack hijack. The Socket will generate a IO and client_io objects and one of the proc called the env[’rack.hijack’] and assign it with the IO object, therefore the typical http connection is upgraded to a bidirectional connection which is our websocket connection. 

The code briefly shows part of the process.

```ruby
#https://github.com/rails/rails/blob/main/actioncable/test/connection/client_socket_test.rb
def open_connection
      env = Rack::MockRequest.env_for "/test",
        "HTTP_CONNECTION" => "upgrade", "HTTP_UPGRADE" => "websocket",
        "HTTP_HOST" => "localhost", "HTTP_ORIGIN" => "http://rubyonrails.com"
      io, client_io = \
        begin
          Socket.pair(Socket::AF_UNIX, Socket::SOCK_STREAM, 0)
        rescue
          StringIO.new
        end
      env["rack.hijack"] = -> { env["rack.hijack_io"] = io }
```

The server returns the 101 switch protocol indicate that the server allowed the connection and upgrade it to the Websocket protocol to the client. we now have a socket connection that is maintained open. And it doesn’t close the connection until the client or server one side close the connection.

Below is what the client will get from the response. 

```ruby
HTTP/1.1 101 Web Socket Protocol Handshake
Upgrade: WebSocket
Connection: Upgrade
WebSocket-Origin: http://localhost:3000
WebSocket-Location: ws://localhost:3000ws://localhost:3000/cable
```

Now the server has accepted our Websocket request, a connection object is created. All channel subscriptions that are established are that connection object is the children of that connection object. The connection itself does nothing but only deal with the authentication and authorization. The client of the websocket connection is called “consumer”. Each browser tab, window or device that an individual user open will create one consumer-connection pair.

Connections are instances of `ApplicationCable::Connection`, which extends `[ActionCable::Connection::Base](https://api.rubyonrails.org/v7.0.2/classes/ActionCable/Connection/Base.html)`. In `ApplicationCable::Connection`, We will authorize the incoming connection and proceed to establish it if the user can be identified. Below shows a typical setup for the connection file.

```ruby
*# app/channels/application_cable/connection.rb*module ApplicationCable
  class **Connection** **<** ActionCable**::**Connection**::**Base
    identified_by :current_user

    def **connect**self.**current_user** **=** find_verified_user
    end

    **private**def **find_verified_user**if verified_user **=** User.**find_by**(id: cookies.**encrypted**[:user_id])
          verified_user
        else
          reject_unauthorized_connection
        end
      end
  end
end

```

Here `[identified_by](https://api.rubyonrails.org/v7.0.2/classes/ActionCable/Connection/Identification/ClassMethods.html#method-i-identified_by)` designates a connection identifier that can be used to find the specific connection later. In code above, we are passing user_id as our http request header every time, Note that anything marked as an identifier will automatically create a delegate by the same name on any channel instances created off the connection. 

```ruby
# File actioncable/lib/action_cable/connection/identification.rb, line 20
def identified_by(*identifiers)
  Array(identifiers).each { |identifier| attr_accessor identifier }
  self.identifiers += identifiers
end
```

This example assumes the user is already authenticated, and has an encrypted cookie with the user ID. The cookie is then automatically sent to the connection instance when a new connection is attempted, and assign to `current_user` variable. By identifying the connection by matching the same current user, we are ensuring that we can later retrieve all open connections by providing with a user (and potentially disconnect them all if the user is deleted or unauthorized).

The official documentation also suggested that if we are using cookies to store for a session then we might retrieve the user_id by using the below approach ( with the assumption of session is named _session and the user id is stored inside a variable called ‘user_id’ )

`verified_user **=** User.**find_by**(id: cookies.**encrypted**['_session']['user_id'])`

Reply:
When do we use HTTP vs websocket?

Reply:
**When can a web socket be used:** 

- **Real-time web application:** A web socket is used in a real-time web application to display data at the client end that is continuously sent by the backend server. WebSocket is faster and enhances application performance since data is continuously pushed/transmitted into the same connection that is already open. For example, in a trading website or bitcoin trading, the backend server continuously pushes price fluctuation and movement data to the client end over a WebSocket channel.
- **Gaming application:** A web socket is used in a gaming application because data is continuously received by the server and will take impact on the screen without refreshing the UI. The UI gets automatically refreshed without even creating a new connection, which is quite useful in a Gaming application.
- **Chat application:** WebSockets are used by chat applications to create a connection just once for the purpose of exchanging, publishing, and broadcasting messages among subscribers. For sending and receiving messages, as well as one-to-one message transfer, it uses the same WebSocket connection.

**When not to use WebSocket:** If we want any real-time updated or continuous streams of data to be communicated over the network, we can utilise WebSocket. If we **want to get old data** or **get it only once** to process it with an application, we should utilise HTTP protocol. Old data that isn't needed very often or fetched only once can be queried with a simple HTTP request, thus we shouldn't use WebSocket in this case.

RESTful web services are sufficient to get the data from the server if we are loading the data only once.

Reply:
So, why not just use websocket for everything?